From: Alan Modra <amodra@gmail.com>
Date: Wed, 21 Jun 2023 22:44:09 +0930
Subject: [PATCH] PR30536, ppc64el gold linker produces unusable clang-16
 binary

In commit 0961e631575b, the fix for PR30217, make_lplt_section and
make_brlt_section were changed to use rela_dyn_ rather than their own
separate dynamic reloc sections.  This fails miserably whenever brlt_
is needed for long branches, due to needing to iterate sizing and thus
reset brlt_ sizes.

	PR 30536
	PR 30217
	* powerpc.cc (Target_powerpc::make_brlt_section): Don't use
	rela_dyn_.

(cherry picked from commit dc808a2ebab337b5517add4c1aad298cf836c239)
---
 gold/powerpc.cc | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/gold/powerpc.cc b/gold/powerpc.cc
index a847dadf9003..aa2d9a187874 100644
--- a/gold/powerpc.cc
+++ b/gold/powerpc.cc
@@ -4632,7 +4632,15 @@ Target_powerpc<size, big_endian>::make_brlt_section(Layout* layout)
 	{
 	  // When PIC we can't fill in .branch_lt but must initialise at
 	  // runtime via dynamic relocations.
-	  brlt_rel = this->rela_dyn_section(layout);
+	  this->rela_dyn_section(layout);
+	  // FIXME: This reloc section won't have its relative relocs
+	  // sorted properly among the other relative relocs in rela_dyn_
+	  // but it must be a separate section due to needing to call
+	  // reset_data_size().
+	  brlt_rel = new Reloc_section(false);
+	  if (this->rela_dyn_->output_section())
+	    this->rela_dyn_->output_section()
+	      ->add_output_section_data(brlt_rel);
 	}
       this->brlt_section_
 	= new Output_data_brlt_powerpc<size, big_endian>(this, brlt_rel);
